### ISSUE TITLE
train with SGD does not work with int32

### ISSUE THREAD

**Issue by julianfrank** [tensorflow/tfjs#898]
To get help from the community, check out our [Google group](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs).

#### TensorFlow.js version
tfjs: "0.13.4"
tfjs-converter: "0.6.6"
tfjs-core: "0.13.10"
tfjs-layers: "0.8.4"
#### Browser version
chrome 70
#### Describe the problem or feature request
I tried to change the example given for SGD based training but for pur int32 constants and variables.... I get the following error which says I'm trying to add a float32 and int32...In my code that is not the case (see below)...
```
tfjs@latest:2 Uncaught Error: The dtypes of the first(float32) and second(int32) input must match
    at assert (tfjs@latest:2)
    at assertTypesMatch (tfjs@latest:2)
    at add_ (tfjs@latest:2)
    at Object.add (tfjs@latest:2)
    at e.add (tfjs@latest:2)
    at tfjs@latest:2
    at tfjs@latest:2
    at e.scopedRun (tfjs@latest:2)
    at e.tidy (tfjs@latest:2)
    at e.tidy (tfjs@latest:2)
```

#### Code to reproduce the bug / link to feature request
```
// Fit a quadratic function by learning the coefficients a, b, c.
const xs = tf.tensor1d([1, 2, 3, 4],"int32");
const ys = tf.tensor1d([2, 4, 6, 8],"int32");
const a = tf.scalar(Math.random(),"int32").variable();
const b = tf.scalar(Math.random(),"int32").variable();
const c = tf.scalar(Math.random(),"int32").variable();
// y = a * x^2 + b * x + c.
const f = x => a.mul(x.square()).add(b.mul(x)).add(c);
const loss = (pred, label) => pred.sub(label).square().mean();
const learningRate = 0.009;
const optimizer = tf.train.sgd(learningRate);
// Train the model.
for (let i = 0; i < 10; i++) {
   optimizer.minimize(() => loss(f(xs), ys));
}
// Make predictions.
console.log(
     `a: ${a.dataSync()}, b: ${b.dataSync()}, c: ${c.dataSync()}`);
const preds = f(xs).dataSync();
preds.forEach((pred, i) => {
   log(`x: ${i}, pred: ${pred}`);
});
```

**Comment by tafsiri** [comment]
@julianfrank Ultimately the issue here is not in your code but stems from how the SGD optimizer works, in order to do its optimization it will in each step try and multiply the variable by some small number (which will be a float) and that is where the error occurs. This means that the variables need to be float32. That however in turn means that the xs and ys should be float32 to be compatible with the loss function you have written (or you can temporarily cast them in the loss function to float32). Hope that helps.

If you were using the codepen link, I've made an issue here to track a fix that will make the stack traces readable so that it is clearer where the error is coming from https://github.com/tensorflow/tfjs/issues/905.

**Comment by julianfrank** [comment]
Thanks for clarifying... Maybe the documentation can explicitly mention that SGD needs float32 as inputs ...
